File: backend/models/book.go
package models

import (
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Book struct {
	Title           string  `gorm:"type:varchar(255);not null" json:"title"`
	ID              string  `gorm:"primaryKey;type:text" json:"id"`
	Description     string  `gorm:"type:text" json:"description"`
	Category        string  `gorm:"type:varchar(100)" json:"category"`
	Image           string  `gorm:"type:varchar(255)" json:"image"`
	PublicationYear int     `gorm:"type:int" json:"publication_year"`
	Rating          float64 `gorm:"type:float" json:"rating"`
	Pages           int     `gorm:"type:int" json:"pages"`
	Isbn            string  `gorm:"type:varchar(20);unique" json:"isbn"`
	CreatedAt       int64   `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt       int64   `gorm:"autoUpdateTime" json:"updated_at"`
	AuthorName      string  `gorm:"type:varchar(100)" json:"author_name"`
}

func (book *Book) BeforeCreate(tx *gorm.DB) error {
	if book.ID == "" {
		book.ID = uuid.New().String()
	}
	return nil
}

File: frontend/src/types/book.ts
import { getBooks } from "@/actions/book.actions";

export enum bookCategory {
  FICTION = "fiction",
  NON_FICTION = "non_fiction",
  SCIENCE = "science",
  HISTORY = "history",
  FANTASY = "fantasy",
  MYSTERY = "mystery",
  THRILLER = "thriller",
  COOKING = "cooking",
  TRAVEL = "travel",
  CLASSICS = "classics",
}

export type Book = {
  id: string;
  title: string;
  description: string;
  author_name: string;
  category: bookCategory;
  image: string;
  publication_year: number;
  rating: number;
  pages: number;
  isbn: string;
  created_at: Date;
  updated_at: Date;
}

export interface Filters {
  publication_year?: string
  category?: string
  query?: string
  sort?: string
  pages?: string
  rating?: string
}


export interface BooksResponse {
  data: Book[]
  meta: {
    total_count: number
    limit: number
    offset: number
  }
}

export type BooksDetails = Awaited<ReturnType<typeof getBooks>>;